uniffi_macros/src/export/scaffolding.rs

```diff
@@ -3,81 +3,169 @@
         &uniffi_meta::fn_ffi_symbol_name(mod_path, &name_s, checksum),
         Span::call_site(),
     );
 
-    let mut params = Vec::new();
-    let mut args = Vec::new();
+    const ERROR_MSG: &str =
+        "uniffi::export must be used on the impl block, not its containing fn's";
+    let (params, args): (Vec<_>, Vec<_>) = collect_params(&sig.inputs, ERROR_MSG).unzip();
 
-    for (i, arg) in item.sig.inputs.iter().enumerate() {
-        match arg {
-            FnArg::Receiver(receiver) => {
-                return Err(syn::Error::new_spanned(
-                    receiver,
-                    "methods are not yet supported by uniffi::export",
-                ));
-            }
-            FnArg::Typed(pat_ty) => {
-                let ty = &pat_ty.ty;
-                let name = format_ident!("arg{i}");
+    let fn_call = quote! {
+        #name(#(#args),*)
+    };
 
-                params.push(quote! { #name: <#ty as ::uniffi::FfiConverter>::FfiType });
+    gen_ffi_function(sig, ffi_ident, &params, fn_call)
+}
 
-                let panic_fmt = match &*pat_ty.pat {
-                    Pat::Ident(i) => {
-                        format!("Failed to convert arg '{}': {{}}", i.ident)
-                    }
-                    _ => {
-                        format!("Failed to convert arg #{i}: {{}}")
-                    }
-                };
-                args.push(quote! {
-                    <#ty as ::uniffi::FfiConverter>::try_lift(#name).unwrap_or_else(|err| {
-                        ::std::panic!(#panic_fmt, err)
-                    })
-                });
+pub(super) fn gen_method_scaffolding(
+    sig: &syn::Signature,
+    mod_path: &[String],
+    checksum: u16,
+    self_ident: &Ident,
+) -> TokenStream {
+    let name = &sig.ident;
+    let name_s = name.to_string();
+
+    let ffi_name = format!("impl_{self_ident}_{name_s}");
+    let ffi_ident = Ident::new(
+        &uniffi_meta::fn_ffi_symbol_name(mod_path, &ffi_name, checksum),
+        Span::call_site(),
+    );
+
+    let mut params_args = (Vec::new(), Vec::new());
+
+    const RECEIVER_ERROR: &str = "unreachable: only first parameter can be method receiver";
+    let fn_call_prefix = match sig.inputs.first() {
+        Some(arg) if is_receiver(arg) => {
+            let ffi_converter = quote! {
+                <::std::sync::Arc<#self_ident> as ::uniffi::FfiConverter>
+            };
+
+            params_args.0.push(quote! { this: #ffi_converter::FfiType });
+
+            let remaining_args = sig.inputs.iter().skip(1);
+            params_args.extend(collect_params(remaining_args, RECEIVER_ERROR));
+
+            quote! {
+                #ffi_converter::try_lift(this).unwrap_or_else(|err| {
+                    ::std::panic!("Failed to convert arg 'self': {}", err)
+                }).
             }
         }
-    }
+        _ => {
+            params_args.extend(collect_params(&sig.inputs, RECEIVER_ERROR));
+            quote! { #self_ident:: }
```

Comment:
It's very nice how this adds support for static functions on the scaffolding side, but I don't think the support is there on the bindings side yet.  Maybe this should be commented out and replaced with a panic for now.

Reply:
Right. Adressed in latest commit.